src/rust/lib/graph-editor/src/component/visualization/renderer/class.rs

```diff
@@ -0,0 +1,69 @@
+//! This module defines the `DataRenderer` trait and related functionality.
+
+use crate::prelude::*;
+use crate::visualization::*;
+
+use crate::frp;
+
+use ensogl::display;
+
+
+
+// ===========
+// === FRP ===
+// ===========
+
+/// FRP api of a `DataRenderer`.
+#[derive(Clone,Debug)]
+#[allow(missing_docs)]
+pub struct DataRendererFrp {
+    pub network   : frp::Network,
+    /// This is emitted if the state of the renderer has been changed by UI interaction.
+    /// It contains the output data of this visualisation if there is some.
+    pub on_change : frp::Source<Option<Data>>,
+}
+
+impl Default for DataRendererFrp {
+    fn default() -> Self {
+        frp::new_network! { renderer_events
+            def on_change = source::<Option<Data>> ();
+        };
+        let network = renderer_events;
+        Self {network,on_change}
+    }
+}
+
+
+
+// ====================
+// === DataRenderer ===
+// ====================
+
+/// At the core of the visualization system sits a `DataRenderer`. The DataRenderer is in charge of
+/// producing a `display::Object` that will be shown in the scene. It will create FRP events to
+/// indicate updates to its output data (e.g., through user interaction).
+///
+/// A DataRenderer can indicate what kind of data it can use to create a visualisation through the
+/// `valid_input_types` method. This serves as a hint, it will also reject invalid input in the
+/// `set_data` method with a `DataError`. The owner of the `DataRenderer` is in charge of producing
+/// UI feedback to indicate a problem with the data.
+pub trait DataRenderer: display::Object + Debug {
+    /// Indicate which `DataType`s can be rendered by this visualization.
+    fn valid_input_types(&self) -> Vec<DataType> {
+        // TODO this will need to be implemented by each Renderer once we get to the registry.
+        unimplemented!()
+    }
+    /// Set the data that should be rendered. If the data is valid, it will return the data as
+    /// processed by this `DataRenderer`, if the data is of an invalid data type, ir violates other
+    /// assumptions of this `DataRenderer`, a `DataError` is returned.
+    fn set_data(&self, data:Data) -> Result<(), DataError>;
+    /// Set the size of viewport of the visualization. The visualisation must not render outside of
+    /// this viewport.
+    fn set_size(&self, size:Vector2<f32>);
```

Comment:
The idea here is, that the higher-level logic should be part of the `Visualization`. So the `DataRenderer` is getting called from the `Visualisation` and we avoid the boilerplate of doing all the FRP interactions on specific errors in each `DataRenderer`.

For example, we get invalid data. Instead of each `DataRenderer` requiring to have an FRP set-up and emitting a signal when there is invalid data, instead the `DataRenderer` just returns the error when it gets called from the `Visualization` and the `Visualization` does all of the work of propagating the correct FRP event. 

This seems to be more ergonomic for the `DataRenderer` implementation since it reduces the complexity of the required behavior. It also more clearly enforces the expected behavior through the type system: a `Result` must be returned, instead of the implicit expectation, that an FRP event should be emitted.

Reply:
I understand that, but this breaks some core assumptions of frp systems. I would still design all interactions between components using frp. Definitely setting data to visualization should be done by frp endpoint. What you have written about security etc of course makes sense, but lets start with frp on such endpoints and lets see over time ifthey are really bad here. In such case, let's talk about it, but for now, lets use frp for all communication.